File: src/lib/game/broadcast.ts
// ============================
// 游戏引擎 - 广播系统
// ============================
import { GameState, Player, Card, BroadcastSubtype, BroadcastResponse } from './types';
import { addLog } from './utils';
import { drawCard } from './deck';
import { getDistance, getSystemsInRange } from './starmap';
import { interruptTurn, resumeTurn } from './turn';

/**
 * 发起广播
 * @returns {boolean} 是否成功创建广播
 */
export function initiateBroadcast(
  state: GameState,
  playerId: string,
  cardUid: string,
  targetSystem: number
): boolean {
  const player = state.players.find(p => p.id === playerId)!;
  const cardIndex = player.hand.findIndex(c => c.uid === cardUid);

  console.log('[initiateBroadcast] 开始创建广播:', {
    playerId,
    playerName: player?.name,
    cardUid,
    cardIndex,
    handSize: player?.hand.length,
    targetSystem,
  });

  if (cardIndex === -1) {
    console.warn('[initiateBroadcast] 卡牌不在手中:', { cardUid });
    return false;
  }
  const card = player.hand[cardIndex];

  if (card.type !== 'broadcast') {
    console.warn('[initiateBroadcast] 不是广播牌:', { cardType: card.type, cardName: card.name });
    return false;
  }
  if (player.energy < card.energy) {
    console.warn('[initiateBroadcast] 能量不足:', { playerEnergy: player.energy, cardEnergy: card.energy });
    return false;
  }

  // 检查广播限制
  const recentBroadcast = player.broadcastHistory.find(
    h => h.systemId === targetSystem && state.totalTurn - h.turn < 2
  );
  if (recentBroadcast) {
    addLog(state, `${player.name} 不能连续在同一星系广播`, 'system');
    console.warn('[initiateBroadcast] 连续广播限制:', {
      targetSystem,
      lastBroadcastTurn: recentBroadcast.turn,
      currentTurn: state.totalTurn,
    });
    return false;
  }

  // 消耗能量,移除手牌
  player.energy -= card.energy;
  player.hand.splice(cardIndex, 1);

  const range = card.range ?? 1;
  player.broadcastHistory.push({ systemId: targetSystem, turn: state.totalTurn });

  // 确定可以回应的玩家
  const responses: BroadcastResponse[] = [];
  for (const other of state.players) {
    if (other.id === playerId || other.eliminated) continue;
    const dist = getDistance(other.position, targetSystem);
    const inRange = dist <= range;

    if (!inRange) continue;

    // 检查是否有合适的广播牌且有足够能量
    // 需要检查: 1) 有广播牌 2) 牌的 range >= 当前广播的 range 3) 有足够能量支付该牌
    const hasValidBroadcastCard = other.hand.some(
      c => c.type === 'broadcast' && (c.range ?? 0) >= range && other.energy >= c.energy
    );
    const hasBroadcastCard = other.hand.some(c => c.type === 'broadcast');
    const hasEnergy = other.energy >= (other.hand.find(c => c.type === 'broadcast')?.energy ?? 0);

    // 检查监听基地 (允许不做回应)
    const hasMonitoringStation = other.faceUpCards.some(c => c.ability === 'detect_broadcast');

    const isAtTarget = other.position === targetSystem;
    // 有监听基地时,即使在目标星系也不必回应
    const mustRespond = isAtTarget && hasValidBroadcastCard && !hasMonitoringStation;

    responses.push({
      playerId: other.id,
      playerName: other.name,
      canRespond: hasValidBroadcastCard,
      mustRespond,
      responded: false,
      agreed: false,
    });
  }

  console.log('[initiateBroadcast] 可能的回应者:', {
    totalPlayers: state.players.length,
    responsesCount: responses.length,
    possibleResponders: responses.filter(r => r.canRespond).map(r => r.playerName),
  });

  state.broadcast = {
    active: true,
    broadcasterId: playerId,
    cardUid,
    card,
    targetSystem,
    range,
    subtype: card.subtype ?? 'cooperation',
    responses,
    phase: 'waiting',
  };

  addLog(state, `${player.name} 向星系 ${targetSystem} 发送了【${card.name}】 (${card.subtype === 'cooperation' ? '合作' : '伪装'})`, 'broadcast');

  // 检查是否有可以回应的玩家
  const possibleResponders = responses.filter(r => r.canRespond);
  if (possibleResponders.length === 0) {
    // 无人回应
    player.energy += 1;
    state.discardPile.push(card);
    addLog(state, `无人回应广播, ${player.name} 获得 1 点能量`, 'broadcast');
    console.log('[initiateBroadcast] 无人可以回应，清理广播状态');
    state.broadcast = null;
  } else {
    // 中断当前回合流程,等待广播响应
    interruptTurn(state, '等待广播响应');
  }

  return true;
}

/**
 * 玩家选择回应广播
 */
export function respondToBroadcast(state: GameState, playerId: string, agreed: boolean, cardUid?: string): void {
  if (!state.broadcast) return;
  const response = state.broadcast.responses.find(r => r.playerId === playerId);
  if (!response) return;

  response.agreed = agreed;
  response.responded = true;

  if (agreed && cardUid) {
    const player = state.players.find(p => p.id === playerId)!;
    const card = player.hand.find(c => c.uid === cardUid);
    if (card) {
      response.responseCard = card;
    }
  }
}

/**
 * 广播发布者选择回应者
 * 注意：这个函数只设置选中的回应者和阶段，不负责结算
 * 结算由 BroadcastFlowManager.executeBroadcastResolution 处理
 */
export function selectBroadcastResponder(state: GameState, responderId: string): void {
  if (!state.broadcast) return;
  state.broadcast.selectedResponderId = responderId;
  state.broadcast.phase = 'reveal';  // 进入揭示阶段
  // 注意：不调用 resolveBroadcast，由 BroadcastFlowManager 控制结算时机
}

/**
 * 结算广播
 */
export function resolveBroadcast(state: GameState): void {
  if (!state.broadcast || !state.broadcast.selectedResponderId) return;

  const broadcaster = state.players.find(p => p.id === state.broadcast!.broadcasterId)!;
  const responder = state.players.find(p => p.id === state.broadcast!.selectedResponderId)!;
  const response = state.broadcast.responses.find(r => r.playerId === state.broadcast!.selectedResponderId)!;

  const bSubtype = state.broadcast.subtype;
  const rSubtype = (response.responseCard?.subtype ?? 'cooperation') as BroadcastSubtype;

  // 消耗回应者的能量和牌
  if (response.responseCard) {
    const cardIdx = responder.hand.findIndex(c => c.uid === response.responseCard!.uid);
    if (cardIdx >= 0) {
      responder.energy -= response.responseCard.energy;
      responder.hand.splice(cardIdx, 1);
    }
  }

  // 结算
  let bEnergy = 0;
  let rEnergy = 0;

  if (bSubtype === 'cooperation' && rSubtype === 'cooperation') {
    bEnergy = 3;
    rEnergy = 3;
    addLog(state, `双方合作! ${broadcaster.name} 和 ${responder.name} 各获得 3 点能量`, 'broadcast');
  } else if (bSubtype === 'disguise' && rSubtype === 'cooperation') {
    bEnergy = 5;
    addLog(state, `${broadcaster.name} 伪装成功! 获得 5 点能量`, 'broadcast');
  } else if (bSubtype === 'cooperation' && rSubtype === 'disguise') {
    rEnergy = 5;
    addLog(state, `${responder.name} 伪装成功! 获得 5 点能量`, 'broadcast');
  } else {
    addLog(state, `双方伪装! 无人获得能量`, 'broadcast');
  }

  broadcaster.energy += bEnergy;
  responder.energy += rEnergy;

  // 回应者补 1 张牌
  const drawn = drawCard(state, 1);
  responder.hand.push(...drawn);

  // 广播牌明牌放在发布者面前
  broadcaster.faceUpCards.push(state.broadcast.card);

  // 清理广播状态
  state.broadcast = null;
  state.pendingAction = null;

  // 恢复中断的回合流程
  resumeTurn(state);

  // 检查游戏是否结束
  const alivePlayers = state.players.filter(p => !p.eliminated);
  if (alivePlayers.length <= 1) {
    state.phase = 'gameOver';
    if (alivePlayers.length === 1) {
      state.winner = alivePlayers[0].id;
    } else {
      state.winner = null;
    }
    return;
  }
}

/**
 * 取消广播 (无人回应时)
 */
export function cancelBroadcast(state: GameState): void {
  if (!state.broadcast) return;
  const player = state.players.find(p => p.id === state.broadcast!.broadcasterId)!;
  player.energy += 1;
  state.discardPile.push(state.broadcast.card);
  addLog(state, `无人回应, ${player.name} 获得 1 点能量`, 'broadcast');
  state.broadcast = null;
  state.pendingAction = null;
  
  // 恢复中断的回合流程
  resumeTurn(state);
}

/**
 * 检查星系是否在广播范围内
 */
export function isSystemInRange(from: number, to: number, range: number): boolean {
  return getDistance(from, to) <= range;
}

/**
 * 获取星系中的存活玩家
 */
export function getPlayersAtSystem(state: GameState, systemId: number): Player[] {
  return state.players.filter(p => !p.eliminated && p.position === systemId);
}

File: backend/internal/game/broadcast.go
package game

import (
	"fmt"
	"slices"
)

func InitiateBroadcast(state *GameState, playerID string, cardUID string, targetSystem int) bool {
	var player *Player
	for i := range state.Players {
		if state.Players[i].ID == playerID {
			player = &state.Players[i]
			break
		}
	}
	if player == nil {
		return false
	}

	cardIndex := slices.IndexFunc(player.Hand, func(c Card) bool { return c.UID == cardUID })
	if cardIndex == -1 {
		return false
	}
	card := player.Hand[cardIndex]

	if card.Type != CardTypeBroadcast {
		return false
	}
	if player.Energy < card.Energy {
		return false
	}

	recentBroadcast := slices.ContainsFunc(player.BroadcastHistory, func(h struct {
		SystemID int
		Turn     int
	}) bool {
		return h.SystemID == targetSystem && state.TotalTurn-h.Turn < 2
	})
	if recentBroadcast {
		return false
	}

	player.Energy -= card.Energy
	player.Hand = append(player.Hand[:cardIndex], player.Hand[cardIndex+1:]...)

	rangeVal := 1
	if card.Range != nil {
		rangeVal = *card.Range
	}
	player.BroadcastHistory = append(player.BroadcastHistory, struct {
		SystemID int
		Turn     int
	}{SystemID: targetSystem, Turn: state.TotalTurn})

	var responses []BroadcastResponse
	for i := range state.Players {
		other := &state.Players[i]
		if other.ID == playerID || other.Eliminated {
			continue
		}
		dist := GetDistance(other.Position, targetSystem)
		if dist > rangeVal {
			continue
		}

		hasValidBroadcastCard := slices.ContainsFunc(other.Hand, func(c Card) bool {
			if c.Type != CardTypeBroadcast {
				return false
			}
			return other.Energy >= c.Energy
		})

		hasMonitoringStation := slices.ContainsFunc(other.FaceUpCards, func(c Card) bool {
			return c.Ability != nil && *c.Ability == "detect_broadcast"
		})

		isAtTarget := other.Position == targetSystem
		mustRespond := isAtTarget && hasValidBroadcastCard && !hasMonitoringStation

		responses = append(responses, BroadcastResponse{
			PlayerID:    other.ID,
			PlayerName:  other.Name,
			CanRespond:  hasValidBroadcastCard,
			MustRespond: mustRespond,
			Responded:   false,
			Agreed:      false,
		})
	}

	subtype := BroadcastSubtypeCooperation
	if card.Subtype != nil {
		subtype = *card.Subtype
	}

	state.Broadcast = &BroadcastState{
		BroadcasterID: playerID,
		CardUID:       cardUID,
		Card:          card,
		TargetSystem:  targetSystem,
		Range:         rangeVal,
		Subtype:       subtype,
		Responses:     responses,
		Phase:         BroadcastPhaseWaiting,
	}

	// 广播涉及广播者与所有候选回应者
	broadcastPlayerIDs := []string{playerID}
	for _, r := range responses {
		broadcastPlayerIDs = append(broadcastPlayerIDs, r.PlayerID)
	}
	AddStructuredLog(state, fmt.Sprintf("%s 向星系 %d 发送了【%s】 (手牌: %d 张)", player.Name, targetSystem, card.Name, len(player.Hand)), LogEntryTypeBroadcast, LogFields{
		SystemID:  &targetSystem,
		CardDefID: &card.DefID,
		PlayerIDs: broadcastPlayerIDs,
	})

	possibleResponders := Filter(responses, func(r BroadcastResponse) bool { return r.CanRespond })
	if len(possibleResponders) == 0 {
		// 设计意图：固定退还 1 点能量以防止刷广播试探，即使原卡牌能量为 2 也仅退 1 点
		player.Energy += 1
		state.DiscardPile = append(state.DiscardPile, card)
		AddStructuredLog(state, fmt.Sprintf("无人回应广播, %s 获得 1 点能量", player.Name), LogEntryTypeBroadcast, LogFields{
			SystemID:  &targetSystem,
			CardDefID: &card.DefID,
			PlayerIDs: []string{playerID},
		})
		state.Broadcast = nil
	} else {
		InterruptTurn(state, "等待广播响应")
	}

	return true
}

func RespondToBroadcast(state *GameState, playerID string, agreed bool, cardUID *string) {
	if state.Broadcast == nil {
		return
	}
	// 入口校验：agreed=true 时 cardUID 必须非空，否则记日志并直接返回（不记录 Agreed=true）
	if agreed && (cardUID == nil || *cardUID == "") {
		AddStructuredLog(state, fmt.Sprintf("回应广播时同意但未提供回应卡: %s", playerID), LogEntryTypeAction, LogFields{
			PlayerIDs: []string{playerID},
		})
		return
	}

	for i := range state.Broadcast.Responses {
		if state.Broadcast.Responses[i].PlayerID == playerID {
			state.Broadcast.Responses[i].Agreed = agreed
			state.Broadcast.Responses[i].Responded = true

			if agreed && cardUID != nil {
				var player *Player
				for j := range state.Players {
					if state.Players[j].ID == playerID {
						player = &state.Players[j]
						break
					}
				}
				if player != nil {
					for _, c := range player.Hand {
						if c.UID == *cardUID {
							state.Broadcast.Responses[i].ResponseCard = &c
							break
						}
					}
				}
			}
			break
		}
	}

	// 所有可回应者都已回应后，推进广播阶段
	allResponded := true
	anyAgreed := false
	for i := range state.Broadcast.Responses {
		r := &state.Broadcast.Responses[i]
		if r.CanRespond && !r.Responded {
			allResponded = false
			continue
		}
		if r.Responded && r.Agreed {
			anyAgreed = true
		}
	}

	if !allResponded {
		return
	}

	if anyAgreed {
		// 至少一人同意：进入选择阶段，等待广播者选择回应者
		state.Broadcast.Phase = BroadcastPhaseSelect
	} else {
		// 无人同意：自动取消广播，退还 1 点能量并恢复回合
		CancelBroadcast(state, state.Broadcast.BroadcasterID)
	}
}

func SelectBroadcastResponder(state *GameState, playerID string, responderID string) {
	if state.Broadcast == nil {
		return
	}
	// 授权校验：仅广播发起者可选择回应者
	if state.Broadcast.BroadcasterID != playerID {
		return
	}
	// 输入校验：responderID 必须对应一个 CanRespond && Responded && Agreed 均为 true 的响应
	valid := false
	for _, r := range state.Broadcast.Responses {
		if r.PlayerID == responderID && r.CanRespond && r.Responded && r.Agreed {
			valid = true
			break
		}
	}
	if !valid {
		AddStructuredLog(state, fmt.Sprintf("无效的回应者选择: %s", responderID), LogEntryTypeAction, LogFields{
			PlayerIDs: []string{responderID},
		})
		return
	}
	state.Broadcast.SelectedResponderID = &responderID
	state.Broadcast.Phase = BroadcastPhaseReveal
	// 选择回应者后立即结算：揭示双方卡牌、结算能量、恢复回合。
	// 否则 Broadcast 永远非 nil、TurnPhase 永远停在 interrupted，广播者卡死。
	ResolveBroadcast(state)
}

func ResolveBroadcast(state *GameState) {
	if state.Broadcast == nil || state.Broadcast.SelectedResponderID == nil {
		return
	}

	var broadcaster, responder *Player
	for i := range state.Players {
		if state.Players[i].ID == state.Broadcast.BroadcasterID {
			broadcaster = &state.Players[i]
		}
		if state.Players[i].ID == *state.Broadcast.SelectedResponderID {
			responder = &state.Players[i]
		}
	}
	if broadcaster == nil || responder == nil {
		return
	}

	var response *BroadcastResponse
	for i := range state.Broadcast.Responses {
		if state.Broadcast.Responses[i].PlayerID == *state.Broadcast.SelectedResponderID {
			response = &state.Broadcast.Responses[i]
			break
		}
	}
	if response == nil {
		return
	}

	bSubtype := state.Broadcast.Subtype
	rSubtype := BroadcastSubtypeCooperation
	rAgreed := response.Agreed
	// 守卫：ResponseCard 为 nil 但 Agreed=true 时强制按拒绝处理，避免无卡进入矩阵结算
	if response.ResponseCard == nil {
		rAgreed = false
	}
	if response.ResponseCard != nil && response.ResponseCard.Subtype != nil {
		rSubtype = *response.ResponseCard.Subtype
	}

	if response.ResponseCard != nil {
		cardIdx := slices.IndexFunc(responder.Hand, func(c Card) bool {
			return c.UID == response.ResponseCard.UID
		})
		if cardIdx >= 0 {
			responder.Energy -= response.ResponseCard.Energy
			responder.Hand = append(responder.Hand[:cardIdx], responder.Hand[cardIdx+1:]...)
		}
	}

	var bEnergy, rEnergy int
	// 广播结算日志统一携带目标星系与广播卡牌 defId
	bcTargetSystem := state.Broadcast.TargetSystem
	bcCardDefID := state.Broadcast.Card.DefID
	if rAgreed {
		switch {
		case bSubtype == BroadcastSubtypeCooperation && rSubtype == BroadcastSubtypeCooperation:
			bEnergy = 3
			rEnergy = 3
			AddStructuredLog(state, fmt.Sprintf("双方合作! %s 和 %s 各获得 3 点能量", broadcaster.Name, responder.Name), LogEntryTypeBroadcast, LogFields{
				SystemID:  &bcTargetSystem,
				CardDefID: &bcCardDefID,
				PlayerIDs: []string{broadcaster.ID, responder.ID},
			})
		case bSubtype == BroadcastSubtypeDisguise && rSubtype == BroadcastSubtypeCooperation:
			bEnergy = 5
			AddStructuredLog(state, fmt.Sprintf("%s 伪装成功! 获得 5 点能量", broadcaster.Name), LogEntryTypeBroadcast, LogFields{
				SystemID:  &bcTargetSystem,
				CardDefID: &bcCardDefID,
				PlayerIDs: []string{broadcaster.ID},
			})
		case bSubtype == BroadcastSubtypeCooperation && rSubtype == BroadcastSubtypeDisguise:
			rEnergy = 5
			AddStructuredLog(state, fmt.Sprintf("%s 伪装成功! 获得 5 点能量", responder.Name), LogEntryTypeBroadcast, LogFields{
				SystemID:  &bcTargetSystem,
				CardDefID: &bcCardDefID,
				PlayerIDs: []string{responder.ID},
			})
		default:
			AddStructuredLog(state, "双方伪装! 无人获得能量", LogEntryTypeBroadcast, LogFields{
				SystemID:  &bcTargetSystem,
				CardDefID: &bcCardDefID,
				PlayerIDs: []string{broadcaster.ID, responder.ID},
			})
		}
	}

	broadcaster.Energy += bEnergy
	responder.Energy += rEnergy

	if rAgreed {
		drawn := DrawCard(state, 1)
		responder.Hand = append(responder.Hand, drawn...)
		// 仅在成功进入矩阵结算时累加广播成功次数（Q2）
		broadcaster.BroadcastSuccessCount++
	}

	state.DiscardPile = append(state.DiscardPile, state.Broadcast.Card)

	state.Broadcast = nil
	state.PendingAction = nil

	ResumeTurn(state)

	alivePlayers := Filter(state.Players, func(p Player) bool { return !p.Eliminated })
	if len(alivePlayers) <= 1 {
		state.Phase = GamePhaseGameOver
		if len(alivePlayers) == 1 {
			state.Winner = &alivePlayers[0].ID
		} else {
			state.Winner = nil
		}
	}
}

func CancelBroadcast(state *GameState, playerID string) {
	if state.Broadcast == nil {
		return
	}
	// 授权校验：仅广播发起者可主动取消
	if state.Broadcast.BroadcasterID != playerID {
		return
	}
	var player *Player
	for i := range state.Players {
		if state.Players[i].ID == state.Broadcast.BroadcasterID {
			player = &state.Players[i]
			break
		}
	}
	if player != nil {
		// 设计意图：固定退还 1 点能量以防止刷广播试探，即使原卡牌能量为 2 也仅退 1 点
		player.Energy += 1
	}
	state.DiscardPile = append(state.DiscardPile, state.Broadcast.Card)
	if player != nil {
		cancelTargetSystem := state.Broadcast.TargetSystem
		cancelCardDefID := state.Broadcast.Card.DefID
		AddStructuredLog(state, fmt.Sprintf("无人回应, %s 获得 1 点能量", player.Name), LogEntryTypeBroadcast, LogFields{
			SystemID:  &cancelTargetSystem,
			CardDefID: &cancelCardDefID,
			PlayerIDs: []string{player.ID},
		})
	}
	state.Broadcast = nil
	state.PendingAction = nil

	ResumeTurn(state)
}

func IsSystemInRange(from, to, rangeDist int) bool {
	return GetDistance(from, to) <= rangeDist
}

func GetPlayersAtSystem(state *GameState, systemID int) []*Player {
	var result []*Player
	for i := range state.Players {
		if !state.Players[i].Eliminated && state.Players[i].Position == systemID {
			result = append(result, &state.Players[i])
		}
	}
	return result
}
